### PR TITLE
basic operators

### PR DESCRIPTION
**Pull request by vpuri3**
- [X] identity
- [X] nulloperator
- [x] diffeqscalar
- [X] diffeqarrayoperator
- [X] factorizeddiffeqarrayoperator

### CODE REVIEW

File: src/operators/basic_operators.jl [hunk 1]
```diff
@@ -1,30 +1,105 @@
 """
 $(TYPEDEF)
 """
-struct DiffEqIdentity{T,N} <: AbstractDiffEqLinearOperator{T} end
+struct DiffEqIdentity{N} <: AbstractDiffEqLinearOperator{Bool} end
```
Comment (xtalax) [inline src/operators/basic_operators.jl:4]: Why limit `T` to `Bool`?

File: src/operators/basic_operators.jl [hunk 2]
```diff
@@ -1,30 +1,105 @@
...
 
-DiffEqIdentity(u) = DiffEqIdentity{eltype(u),length(u)}()
-Base.size(::DiffEqIdentity{T,N}) where {T,N} = (N,N)
-Base.size(::DiffEqIdentity{T,N}, m::Integer) where {T,N} = (m == 1 || m == 2) ? N : 1
-LinearAlgebra.opnorm(::DiffEqIdentity{T,N}, p::Real=2) where {T,N} = one(T)
-Base.convert(::Type{AbstractMatrix}, ::DiffEqIdentity{T,N}) where {T,N} =
-                                              LinearAlgebra.Diagonal(ones(T,N))
+# constructors
+DiffEqIdentity(u::AbstractVector) = DiffEqIdentity{length(u)}()
```
Comment (xtalax) [inline src/operators/basic_operators.jl:7]: That's not a `Bool`...

File: src/operators/basic_operators.jl [hunk 3]
```diff
@@ -1,30 +1,105 @@
...
+    @assert issquare(A)
+    N = size(A, 1)
+    DiffEqNullOperator{N}()
 end
-LinearAlgebra.mul!(Y::AbstractVecOrMat, ::DiffEqIdentity, B::AbstractVecOrMat) = Y .= B
-LinearAlgebra.ldiv!(Y::AbstractVecOrMat, ::DiffEqIdentity, B::AbstractVecOrMat) = Y .= B
 
-LinearAlgebra.mul!(Y::AbstractArray, ::DiffEqIdentity, B::AbstractArray) = Y .= B
-LinearAlgebra.ldiv!(Y::AbstractArray, ::DiffEqIdentity, B::AbstractArray) = Y .= B
+Base.convert(::Type{AbstractMatrix}, ::DiffEqNullOperator{N}) where {N} = Diagonal(zeros(Bool, N))
```
Comment (xtalax) [inline src/operators/basic_operators.jl:70]: This may be better as a `DiffEqScalar`

File: src/operators/common_defaults.jl [hunk 4]
```diff
@@ -3,38 +3,28 @@ DEFAULT_UPDATE_FUNC(A,u,p,t) = A # no-op used by the basic operators
 # isconstant(::AbstractDiffEqLinearOperator) = true # already defined in DiffEqBase
 update_coefficients!(L::AbstractDiffEqLinearOperator,u,p,t) = L
 
+Base.size(A::AbstractDiffEqOperator, d::Integer) = d <= 2 ? size(A)[d] : 1
```
Comment (xtalax) [inline src/operators/common_defaults.jl:6]: `size(A, d)` may be better

**Review by xtalax — COMMENTED** [review]
All in all, this is a step forward.

File: src/operators/basic_operators.jl (hunk 3 above)
Comment (xtalax) [reply src/operators/basic_operators.jl:70]: i.e specialise this operator as a scalar

**Comment by vpuri3** [comment]
tests passing

**Pull request merged**